### ISSUE TITLE
USO_UNSAFE_ACCESSIBLE_OBJECT_SYNCHRONIZATION false positive when a static factory sets a lock on a newly allocated instance

### ISSUE THREAD

**Issue by ethanstoner** [spotbugs/spotbugs#4331]
Follow-up to #4320 / #4326, at @JuditKnoll's suggestion on that PR.

While answering the review question about whether the `clone()` fix should be broadened, I hit a second false positive on the same detector. It's a different code path and needs its own test cases, so I kept it out of #4326.

## Reproducer

```java
public class Foo {
    private Object lock = new Object();

    public void doStuff() {
        synchronized (lock) { /* ... */ }
    }

    public static Foo create() {
        Foo f = new Foo();
        f.lock = new Object();   // USO_UNSAFE_ACCESSIBLE_OBJECT_SYNCHRONIZATION reported here
        return f;
    }
}
```

The assignment is safe for the same reason the `clone()` case is: the target instance hasn't escaped yet, so there is no live lock any other thread could be holding.

## Why #4326 doesn't already cover it

#4326 keys on the method being named `clone`, and it has to. `super.clone()` returns from an INVOKESPECIAL, so its result is never `NEWLY_ALLOCATED` on the opcode stack, which puts the `clone()` case out of reach of an `isNewlyAllocated()` check. In a static factory the opposite holds: the *target* is newly allocated. Instrumenting `isNewLockAssignedToCloneCopy` shows both:

```
PROBE method=clone        targetNewlyAllocated=false targetReg=1
PROBE method=create       targetNewlyAllocated=true  targetReg=0   // static factory
PROBE method=resetLockOf  targetNewlyAllocated=false targetReg=1
PROBE method=updateLock4  targetNewlyAllocated=false targetReg=0
```

## Suggested fix

An additional arm in `FindImproperSynchronization.sawOpcode`, PUTFIELD branch: skip `declaredLockAccessors.add` when `stack.getStackItem(1).isNewlyAllocated()`, i.e. when the target is provably unescaped.

## What it must not break

- `UnsafeSynchronizationWithLockExposedInDeclaringClass.updateLock4()` (`lock4 = new Object()` in a public method) has to stay a true positive. A newly allocated *value* is never sufficient; this arm keys on the newly allocated *target*.
- `resetLockOf(Foo other) { other.lock = new Object(); }` has to stay a true positive, since `other` is a parameter and so isn't newly allocated.

Happy to open the PR with test cases once #4326 lands.

Generated with Claude Code on behalf of @ethanstoner.
